### DISCUSSION TITLE
How to define command "ButtonTopic 1" into the user_config_override.h?

### DISCUSSION THREAD

**Discussion by eikejenn** [arendst/Tasmota#20290] [Q&A] [answered]
Hi, 
iam using the latest stable release (13-3-0) and want to configure all my settings by the user_config_override.h.

In the console i can set ButtonTopic to the value "1"
```
ButtonTopic
	<value> = set MQTT button topic
0 = disable use of MQTT button topic
1 = set MQTT button topic to device %topic%
`````
After doing this, tasmota starts sending mqtt messages after pressing buttons. all fine :-)

I want to add this setting into the user override file.
i can find this default entry:
```
#define MQTT_BUTTON_TOPIC      "0"               // [ButtonTopic] MQTT button topic, "0" = same as MQTT_TOPIC, set to 'PROJECT "_BTN_%06X"' for unique topic including device MAC address
```
Changing this to "1" is not the same as the console action. This enables sending messages to to a topic called "1", but not the topic "%topic%", which is set by "#define MQTT_TOPIC"

Second try i tested 
```
#define MQTT_BUTTON_TOPIC      1
```
without the quotation marks. But this runs in a compile error.

Has anyone an idea how to set this option into the firmware file?
Regards 
Eike

**Comment by sfromis** [comment]
Well, do not use the value "1", as this is clearly outside of the comment text guiding what you can use.
> MQTT button topic, "0" = same as MQTT_TOPIC, set to 'PROJECT "_BTN_%06X"' for unique topic including device MAC address

What you do is not directly related to the command, where the value "1" is merely a convenience feature, but not having a special status.

**Comment by eikejenn** [comment]
Hi sfromis, 
thank you for your answer. My goal is to flash a device with the self compiled firmware and then change only the Topic, 
where the device is working with (MQTT_TOPIC) in the WebUi. Changes as little effort as possible.
So how can i directly enable the BUTTON mqtt-messages (in the compiled firmware) to the same Topic as configured in the web ui. 
Exactly whats happen, when i configure "ButtonTopic 1" in the console (1 = set MQTT button topic to device %topic%)

Is the any way?


Regards 
Eike

**Reply by sfromis** [reply]
Well, you can check yourself by seeing what topic you get using that command.

**Reply by eikejenn** [reply]
The goal is not to use the command in the console. Only the main MQTT Topic should be changed manually.
The button topic should AUTOMATICALLY changed to the same topic.
Like an executed "ButtonTopic 1" automatically in the image.

**Reply by barbudor** [reply]
As far as I see the code doesn't allow that.
ButtonTopic command doesn't link the button topic to the general topic permanently
It copies the current value of the general topic to the button topic at the moment the command is executed
For example if you change Topic after a ButtonTopic 1, the button topic will stick to the previous value.

There is no other way than including the command in your provisionning process just like Topic

**Reply by eikejenn** [reply]
Hi bardudor, 
i understand, thanks for the explanation.
Do you see any workaround for me? Can i execute "ButtonTopic 1" at the start of tasmota?
When i change the general topic, the device will restart automatically.  
Can you show me an approach? perhaps i can add one or two lines code?

**Reply by barbudor** [reply]
You should be able to hard code a rule in your binary that is triggered  by the result of the topic command so that each time the topic is set, the button topic will follow

Can't test right now but something like below should be close
```
On topic#data do ButtonTopic 1 endon
```

**Reply by eikejenn** [reply, marked as answer]
Very great thanks :-)
if you know how to do it, it's easy... ^^

your approach 
```
On topic#data do ButtonTopic 1 endon
```
doesnt work, because at the time of sending the "topic" mqtt mesage, the value MQTT_TOPIC is not updated, so that "buttonTopic 1" will set the old topic. :-( Could this be interpreted as a bug? The mqtt message should only be sent when the variable is updated and not the other way around.

I solved this for me with this rule:
```
on topic#data do ButtonTopic %value% endon
```
In this case the ButtonTopic is executed with the value from the received (new) MQTT Topic. This works fine. :-)
perfect.
Thanks again and have nive xmas days.
Regards
Eike
